### PR TITLE
Add better tracking of Collab window

### LINKED ISSUE #9683: Duration in "MsnStatistics" event is tracking garbage
Query:

union Office_Fluid_FluidRuntime_*
| where Data_eventName == "fluid:telemetry:OpPerf:MsnStatistics"
| summarize count(), toint(avg(Data_duration)) by Data_runtimeVersion
| where count_ > 1000
| order by Data_runtimeVersion desc

Numbers are in millions!

Here is the code:
```ts
        if (sequenceNumber % 1000 === 0) {
            if (this.opSendTimeForLatencyStatisticsForMsnStatistics !== undefined) {
                this.logger.sendPerformanceEvent({
                    eventName: "MsnStatistics",
                    sequenceNumber,
                    msnDistance: this.deltaManager.lastSequenceNumber - this.deltaManager.minimumSequenceNumber,
                    duration: message.timestamp - this.opSendTimeForLatencyStatisticsForMsnStatistics,
                });
            }
            this.opSendTimeForLatencyStatisticsForMsnStatistics = message.timestamp;
        }
```

Looking at it, the time basically tracks how long it takes to advance 1000 ops.

### PR DESCRIPTION
**Pull request by vladsud**
Fixes #9683: Duration in "MsnStatistics" event is tracking garbage
Helps better understand #9682: MSN window is too large!

This change adds
- tracking of collab window at summarization, to better understand impact on quality of summary (search, size of catchup blobs).
- tracking of batch sizes, time to inbound and msn size at that time
   - I expect to see msn size correlating with batch sizes, as collab window grows to at least the size of batch at the moment when batch is sent
- Correct tracking of how long it takes for a sequence number to fall out of collab window.

### CODE REVIEW

File: packages/runtime/container-runtime/src/connectionTelemetry.ts [hunk 1]
```diff
@@ -160,16 +162,19 @@ class OpPerfTelemetry {
...
-                    duration: message.timestamp - this.opSendTimeForLatencyStatisticsForMsnStatistics,
-                });
-            }
-            this.opSendTimeForLatencyStatisticsForMsnStatistics = message.timestamp;
+        if (this.sequenceNumberForMsnTracking === undefined && sequenceNumber % 1000 === 0) {
+            this.sequenceNumberForMsnTracking = sequenceNumber;
+            this.msnTrackingTimestamp = message.timestamp;
+        }
+        if (this.sequenceNumberForMsnTracking !== undefined &&
+                message.minimumSequenceNumber >= this.sequenceNumberForMsnTracking) {
```
Comment (NicholasCouri) [inline packages/runtime/container-runtime/src/connectionTelemetry.ts:170]: >message.minimumSequenceNumber >= this.sequenceNumberForMsnTracking [](http://example.com/codeflow?start=16&length=66)

Out of curiosity, can there ever be a scenario in which message.minimumSequenceNumber == message.sequenceNumber  && sequenceNumber %1000 == 0 ?

**Review by NicholasCouri — APPROVED** [review]
:shipit:

File: packages/runtime/container-runtime/src/connectionTelemetry.ts (hunk 1 above)
Comment (vladsud) [reply packages/runtime/container-runtime/src/connectionTelemetry.ts:170]: It should never happen for client ops. Basically, msn tells every client (including future clients who will join) the minimum reference sequence number of future ops. This client (who sent this op) may still sent ops with previous sequence number as a reference sequence number, so MSN can't move yet.

I think the only place where that might be violated is when there are no clients connected to service (i.e. ops are produced by service). last leaveOp or noClient op is the only op I could think of.

I glanced at one file, and it's actually true for both :) I also found summaryAck to have same property (also sent by service). In all these cases clientId == null.

I think it's fine to leave it as is, as we record the truth, even if collab window is empty.

File: packages/runtime/container-runtime/src/connectionTelemetry.ts [hunk 2]
```diff
@@ -160,16 +162,19 @@ class OpPerfTelemetry {
...
+            this.logger.sendPerformanceEvent({
+                eventName: "MsnStatistics",
+                sequenceNumber,
+                msnDistance: sequenceNumber - this.sequenceNumberForMsnTracking,
```
Comment (pragya91) [inline packages/runtime/container-runtime/src/connectionTelemetry.ts:174]: @vladsud -> something does not seem right in the `msnDistance` logic here. My understanding is that we are trying to check the distance of the current client from the MSN from current message (based on the property name `msnDistance`. I think you are trying to sample logs with the help of this.sequenceNumberForMsnTracking, but it cannot be used to calculate msnDistance.

If you are trying to calculate how much time it took to go past the collab window, then maybe we need to rename this variable from `msnDistance` to something else, and add code comments.

**Pull request merged**